allennlp/modules/conditional_random_field.py

```diff
@@ -0,0 +1,170 @@
+"""
+
+import torch
+
+def log_sum_exp(x: torch.Tensor) -> torch.Tensor:  # pylint: disable=invalid-name
+    """
+    Computes log(sum(exp(x))) in a numerically stable way,
+    where the sum is only over the last dimension.
+    """
+    maxes, _ = torch.max(x, -1, keepdim=True)
+    exps = torch.exp(x - maxes)
+
+    return maxes.squeeze(-1) + torch.log(torch.sum(exps, -1))
+
+
+class ConditionalRandomField(torch.nn.Module):
+    """
+    This module uses the "forward-backward" algorithm to compute
+    the log-likelihood of its inputs assuming a conditional random field model.
+
+    See, e.g. http://www.cs.columbia.edu/~mcollins/fb.pdf
+
+    Parameters
+    ----------
+    num_tags: int, required
+        The number of tags.
+    start_tag: int, required
+        The id of the sentinel <START> tag.
+    stop_tag: int, required
+        The id of the sentinel <STOP> tag.
+    """
+    def __init__(self,
+                 num_tags: int,
+                 start_tag: int,
+                 stop_tag: int) -> None:
+        super().__init__()
+
+        self.num_tags = num_tags
+        self.start_tag = start_tag
+        self.stop_tag = stop_tag
+
+        # transitions[i, j] is the logit for transitioning to state i from state j
+        self.transitions = torch.nn.Parameter(torch.randn(num_tags, num_tags))
+
+        # We never transition to the start tag and we never transition from the stop tag.
+        self.transitions.data[start_tag, :] = -10000
+        self.transitions.data[:, stop_tag] = -10000
+
+    def _input_likelihood(self, inputs: torch.Tensor, mask: torch.Tensor) -> torch.Tensor:
+        """
+        Computes the (batch_size,) numerator term for the log-likelihood, which is the
+        sum of the likelihoods across all possible state sequences.
+        """
+        batch_size, sequence_length, num_tags = inputs.size()
+        mask = mask.float()
+        # masks[i] is the (batch_size,) mask tensor for timestep i
+        masks = [mask[:, i].contiguous() for i in range(sequence_length)]
+
+        # We will step through time and accumulate the likelihood at each step.
+        # Note that we will accumulate a (batch_size, num_tags) tensor of likelihoods
+        # since the likelihood of each step is
+
+        # Initial alpha is the (batch_size, num_tags) tensor of likelihoods combining the
+        # transitions to the initial states and the logits for the first timestep.
+        alpha = self.transitions[:, self.start_tag].contiguous().view(1, num_tags) + inputs[:, 0, :]
+
+        # For each i we compute logits for the transitions from timestep i-1 to timestep i
+        # We do so in a (batch_size, num_tags, num_tags) tensor where the axes are
+        # (instance, next_tag, prev_tag)
+        for i in range(1, sequence_length):
+            # The emit scores are for time i ("next_tag") so we broadcast along the prev_tag axis.
+            emit_scores = inputs[:, i].contiguous().view(batch_size, num_tags, 1)
```

Comment:
I typically prefer `unsqueeze` when that's what you're doing, because it's more obvious what the point of the line is.  With `.view` I have to wonder if you've changed any of the other dimensions; with `.unsqueeze` that's not even _possible_, so it's also less prone to errors.  If you like this better, though, I'm fine with you keeping it.